presto-main/src/main/java/com/facebook/presto/operator/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,432 @@
+    private final boolean[] remainingNulls = new boolean[BITS_IN_BYTE];
+
+    // Number of null values not put into nullsBuffer from last batch
+    private int remainingNullsCount;
+
+    // Boolean indicating whether there are any null values in the nullsBuffer. It is possible that all values are non-null.
+    private boolean hasEncodedNulls;
+
+    public static BlockEncodingBuffers createBlockEncodingBuffers(DecodedBlockNode decodedBlockNode)
+    {
+        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
+
+        Object decodedBlock = decodedBlockNode.getDecodedBlock();
+
+        // Skip the Dictionary/Rle block node. The mapping info is not needed when creating buffers.
+        if (decodedBlock instanceof DictionaryBlock) {
+            decodedBlockNode = decodedBlockNode.getChildren().get(0);
+            decodedBlock = decodedBlockNode.getDecodedBlock();
+        }
+        else if (decodedBlock instanceof RunLengthEncodedBlock) {
+            decodedBlockNode = decodedBlockNode.getChildren().get(0);
+            decodedBlock = decodedBlockNode.getDecodedBlock();
+        }
+
+        verify(!(decodedBlock instanceof DictionaryBlock), "Nested RLEs and dictionaries are not supported");
+        verify(!(decodedBlock instanceof RunLengthEncodedBlock), "Nested RLEs and dictionaries are not supported");
+
+        if (decodedBlock instanceof LongArrayBlock) {
+            return new LongArrayBlockEncodingBuffers();
+        }
+
+        if (decodedBlock instanceof Int128ArrayBlock) {
+            return new Int128ArrayBlockEncodingBuffers();
+        }
+
+        if (decodedBlock instanceof IntArrayBlock) {
+            return new IntArrayBlockEncodingBuffers();
+        }
+
+        if (decodedBlock instanceof ShortArrayBlock) {
+            return new ShortArrayBlockEncodingBuffers();
+        }
+
+        if (decodedBlock instanceof ByteArrayBlock) {
+            return new ByteArrayBlockEncodingBuffers();
+        }
+
+        if (decodedBlock instanceof VariableWidthBlock) {
+            return new VariableWidthBlockEncodingBuffers();
+        }
+
+        if (decodedBlock instanceof ColumnarArray) {
+            return new ArrayBlockEncodingBuffers(decodedBlockNode);
+        }
+
+        if (decodedBlock instanceof ColumnarMap) {
+            return new MapBlockEncodingBuffers(decodedBlockNode);
+        }
+
+        if (decodedBlock instanceof ColumnarRow) {
+            return new RowBlockEncodingBuffers(decodedBlockNode);
+        }
+
+        throw new IllegalArgumentException("Unsupported encoding: " + decodedBlock.getClass().getSimpleName());
+    }
+
+    public abstract void resetBuffers();
+
+    public void setNextBatch(int positionsOffset, int batchSize)
+    {
+        this.positionsOffset = positionsOffset;
+        this.batchSize = batchSize;
+    }
+
+    public abstract void copyValues();
```

Comment:
What about `copyValuesForBatch` (any better name is also welcome), and maybe add comments for both `setNextBatch` and this method? --  Otherwise I have to look into `OptimizedPartitionedOutputOperator` to understand the purpose of these two methods.

Reply:
Hi Wenlei, thanks for the suggestion. I renamed it to copyValuesInBatch().